feat: more Realtime docs (#12424)
supabase/supabase: apps/docs/pages/guides/realtime/architecture.mdx

import Layout from '~/layouts/DefaultGuideLayout'

export const meta = {
  id: 'architecture',
  title: 'Realtime Architecture',
  description: 'Architecture of the Supabase Realtime service',
  sidebar_label: 'Architecture',
}

Realtime is a globally distributed Erlang cluster. Clients can connect to any node in the cluster via WebSockets and send messages to any other client connected to the cluster.

> Realtime is written in [Elixir](https://elixir-lang.org/) (a flavor of Erlang) and utilizes many tools the [Phoenix Framework](https://www.phoenixframework.org/) provides out of the box.

Author: > which is interoperable with Erlang and runs on the Erlang virtual machine (BEAM)

I'd just say `which compiles to Erlang`. Saying `interoperable` sounds like it's something totally different. It is Erlang after it's compiled.

> Realtime is written in [Elixir](https://elixir-lang.org/) (a flavor of Erlang) and utilizes many tools the [Phoenix Framework](https://www.phoenixframework.org/) provides out of the box.

Author: Updated, and linked to erlang.org.

Thanks to the Realtime cluster, you (an amazing Supabase dev) don't have to think about which region your users connect to.

If you're using Broadcast, Presence, or streaming database changes, messages will always get to your users via the shortest path possible.

## Connecting to a Database

Realtime allows you to listen to changes from your Postgres database. When a new client connects to Realtime and initializes the `postgres_changes` Realtime Extentsion the cluster will connect to your Postgres database and start streaming changes from a replication slot.

Realtime knows the region your database is in, and connects to it from the closest region possible.

Every Realtime region has at least two nodes so if one node goes offline the other node should reconnect and start streaming changes again.

## Streaming the Write-Ahead Log

During the connection process a Postgres replication slot is acquired.

> Realtime delivers changes by polling the replication slot and appending connection IDs to each wal record. Each connection ID is a WebSocket connection.

Reviewer: So every WebSocket connection has at least one channel and every channel generates a UUID. So a connection can have multiple IDs. Perhaps it's better to call it channel subscription IDs.

```suggestion
Realtime delivers changes by polling the replication slot and appending channel subscription IDs to each wal record.
```

> Connection IDs are Erlang processes representing underlying sockets on the cluster. These IDs are globally unique and messages to processes are routed automatically by the Erlang virtual machine.

Reviewer: Again, probably want to rename `Connection IDs` and maybe stick to calling it Elixir when it's necessary to reference Erlang like in Erlang virtual machine.

Author: Yeah "subscription ID" is better.
